[ZEPPELIN-2411] Improve Table

1ambda opened the pull request:
### What is this PR for?

**Improve Table**

- column filter
- persist column state: type, order, hide/show, sorting, pinning
- pagination
- `setting` menu to configure table UI
- group by + aggregation

And **all these things are persisted and synchronized among web socket clients**

See the screenshot section for more detail.

### What type of PR is it?
[Improvement]

### Todos
* [x] - Remove handsontable dependencies
* [x] - Use npm packaged moment* packages.
* [x] - Apply ui-grid
* [x] - Add setting menu
* [x] - Fix some issues
* [x] - Persist column type

### What is the Jira issue?

[ZEPPELIN-2411](https://issues.apache.org/jira/browse/ZEPPELIN-2411)

### How should this be tested?

1. Build: `mvn clean package -DskipTests; ./bin/zeppelin-daemon.sh restart`
2. Open a note and create tables. If you don't have proper paragraphs, use this snippet.

```scala
%spark

import org.apache.commons.io.IOUtils
import java.net.URL
import java.nio.charset.Charset

val bankText = sc.parallelize(
    IOUtils.toString(
        new URL("https://s3.amazonaws.com/apache-zeppelin/tutorial/bank/bank.csv"),
        Charset.forName("utf8")).split("\n"))

case class Bank(
    age: Integer, 
    job: String, 
    marital: String, 
    education: String, 
    balance: Integer,
    housing: Boolean,
    loan: Boolean,
    contact: String,
    day: Int,
    month: String,
    duration: Int,
    y: Boolean
)

val bank = bankText.map(s => s.split(";")).filter(s => s(0) != "\"age\"").map(
    s => Bank(s(0).toInt, 
            s(1).replaceAll("\"", ""),
            s(2).replaceAll("\"", ""),
            s(3).replaceAll("\"", ""),
            s(5).replaceAll("\"", "").toInt,
            if (s(6).replaceAll("\"", "") == "yes") true else false,
            if (s(7).replaceAll("\"", "") == "yes") true else false,
            s(8).replaceAll("\"", ""),
            s(9).replaceAll("\"", "").toInt,
            s(10).replaceAll("\"", ""),
            s(11).replaceAll("\"", "").toInt,
            if (s(16).replaceAll("\"", "") == "yes") true else false
        )
).toDF()
bank.registerTempTable("bank")
```

```sql
select age, education, job, balance from bank limit 1000
```

### Screenshots (if appropriate)

#### Before

![image](https://cloud.githubusercontent.com/assets/4968473/25803644/d0d81524-3432-11e7-8cf6-dde16465a447.png)

#### After: Filter

![2411_new_filter](https://cloud.githubusercontent.com/assets/4968473/25880089/4e1e6a5e-3570-11e7-8339-36e712a170cc.gif)

#### After: Column related features

![2411_new_column](https://cloud.githubusercontent.com/assets/4968473/25880690/d8a4da02-3573-11e7-9851-cc359fd6075c.gif)

#### After: Pagination

![2411_new_pagination](https://cloud.githubusercontent.com/assets/4968473/25880646/ad3b5ea4-3573-11e7-8e03-e56af683eb9b.gif)

#### After: Group, Aggregation

![2411_new_group_aggr](https://cloud.githubusercontent.com/assets/4968473/25880819/97ba7a82-3574-11e7-8dc4-8d39df6a65a0.gif)

#### After: Synchronized

![2411_new_sync](https://cloud.githubusercontent.com/assets/4968473/25880669/c1784bb6-3573-11e7-966d-cb21d222d683.gif)



### Questions:
* Does the licenses files need update? - YES, updated
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - NO

Leemoonsoo commented:
Tested and overall it's great improvement!

Here's few feedbacks.

1. Table does not refreshed after height / width of paragraph change.
2. How about remove 'type' in column table option?
![image](https://cloud.githubusercontent.com/assets/1540981/25900186/94b065a4-3560-11e7-9491-4700e4228f77.png)
I guess user will be okay without knowing it.
3. showColumnFilter
Following is screenshot with 'showColumnFilter' on. I don't know what this option supposed to do except for adding empty sticky row in the bottom.
![image](https://cloud.githubusercontent.com/assets/1540981/25900262/de274108-3560-11e7-96fb-b5a25a1298ef.png)

4. pagenation
`defaultPaginationSize ` size can be only changed after the Table is displayed with `defaultPaginationSize ` applied. So changing default does not mean too much in this case. Isn't it?

`availablePaginationSizes` - It's minor opinion, I think it is okay with hardcoded value and not make user configurable.

5. Header bar in the option
![image](https://cloud.githubusercontent.com/assets/1540981/25900541/c5bd7d16-3561-11e7-9011-ae61378d39c5.png)

Two buttons - 'Save', 'Delete' icon - on right little bit confuses me. It was difficult to guess what those button supposed to do. And figured out

'save' - apply option to the table.
'delete' - restore to default options

While option change applies immediately, how about hide 'Save' button? And change 'delete' icon to something more appropriate? For example "Restore the default values" text would be more intuitive.

1ambda commented:
Here's few feedbacks.

1. Table does not refreshed after height / width of paragraph change.

TBD

2. How about remove 'type' in column table option?

- Removed
- While adding tooltip for each option

![image](https://cloud.githubusercontent.com/assets/4968473/25916322/2ee4f696-35ff-11e7-806f-b31870e99fb3.png)

3. showColumnFilter

It's used to display aggregated values. I changed the name to `showAggregationFooter` which easier to understand than before.

![image](https://cloud.githubusercontent.com/assets/4968473/25916251/f50ef58e-35fe-11e7-84b0-d784307b1f09.png)

4. pagination

- Removed pagination related settings.

![image](https://cloud.githubusercontent.com/assets/4968473/25916292/168b7bc4-35ff-11e7-8d3f-d58dedad5aae.png)

5. Header bar in the option

- Removed the `save` button.
- Updated icon for reset and added tooltip for it.

![image](https://cloud.githubusercontent.com/assets/4968473/25916335/38401202-35ff-11e7-8712-218145675832.png)

guptarajat commented:
"all these things are persisted and synchronized among web socket clients"
What does this mean?

1ambda commented:
@guptarajat Hi, refer the last GIF. "persist" means that all state of table is saved in the note.josn. For example, sorting state will be saved after this PR but it is not in current master.

Leemoonsoo commented:
Each cell in table supposed to print html. For example

```
%sh echo -e "%table key\tvalue\nsun\t%html <img width=50px src=http://www.freeiconspng.com/uploads/sun-icon-4.jpg></img>"
```

This PR
![image](https://cloud.githubusercontent.com/assets/1540981/26033548/9a728084-387c-11e7-94cf-aa87d16a509d.png)

In master branch
![image](https://cloud.githubusercontent.com/assets/1540981/26033549/a2ca4cda-387c-11e7-905c-3a380b9db4e3.png)

Leemoonsoo commented:
And this PR adds background in rows. 

Master branch
![image](https://cloud.githubusercontent.com/assets/1540981/26033572/048722ea-387d-11e7-99da-1f940daa4347.png)

This PR
![image](https://cloud.githubusercontent.com/assets/1540981/26033584/3ae34ac6-387d-11e7-9b3f-16d1a02eb09f.png)

What do you think? Keep appearance the same as before, so we can keep user experience change minimum. Or is adding background color in rows brings more benefit?

1ambda commented:
@Leemoonsoo 

Colorizing (even rows) is cognitively better because we are not confused when investigating closer rows.

Here are more bright versions

<img width="804" alt="2411_alter" src="https://cloud.githubusercontent.com/assets/4968473/26037441/6eef6ace-392d-11e7-992d-4fb38186852b.png">

<img width="801" alt="2411_white" src="https://cloud.githubusercontent.com/assets/4968473/26037442/764ea582-392d-11e7-9ae8-254b8b3100e3.png">

Leemoonsoo commented:
Bright version looks better for me

1ambda commented:
@Leemoonsoo 

Oops. I missed that feature.

1. Now table renders HTML display in the cells

![image](https://cloud.githubusercontent.com/assets/4968473/26039691/f7bfe146-3958-11e7-83b4-b1f54436e1fc.png)
 
2. Fixed to have more bright even rows.

![image](https://cloud.githubusercontent.com/assets/4968473/26039714/2766e296-3959-11e7-8b01-e82ba277deb9.png)

1ambda commented:
@Leemoonsoo just rebased and

updated to use `white` color in each cell background in c56edca
now it looks like 

![image](https://cloud.githubusercontent.com/assets/4968473/26231787/6a310e9a-3c8c-11e7-8856-f1ac6855c3ef.png)

Leemoonsoo commented:
GTM. Thanks @1ambda for great improvement.

Merge to master if no further discussions.

vglushak commented on zeppelin-web/src/app/visualization/builtins/visualization-table.js:
```diff
@@ -14,50 +14,391 @@
+      // create, compile and append grid elem
+      gridElem = angular.element(
+        `<div id="${gridElemId}" ui-grid="${gridElemId}"
+              ui-grid-edit ui-grid-row-edit 
```
This made all values in the table editable. Is that expected?

vglushak commented:
> Now table renders HTML display in the cells and adjusts row height based on the content.

There are some limitations for this feature. img, links works. But when I want to place button it just ignore html and display text only.

Tagar commented:
Great improvement. 
Submitted https://issues.apache.org/jira/browse/ZEPPELIN-3251 to consider showing rows/columns lazily on scrolling events as otherwise ui-grid has a significant overhead for wide datasets so it randers much slower too. Other ideas?
